### PR TITLE
Support union types

### PR DESCRIPTION
**Pull request by staabm**
before this PR, the extension did not realize methods were called, in case the caller was of a union-type

### CODE REVIEW

**Comment by staabm** [comment]
I have no idea yet how this needs to be fixed

**Comment by staabm** [comment]
I guess we need to return several `MethodCallReference`s from 

https://github.com/TomasVotruba/unused-public/blob/560ec577f67c3a6f6b1530f2a91aa64d0acaae75/src/ClassMethodCallReferenceResolver.php#L38-L46

the underlying issue is related to https://phpstan.org/blog/why-is-instanceof-type-wrong-and-getting-deprecated

**Comment by staabm** [comment]
test fails as expected now:
```
1) TomasVotruba\UnusedPublic\Tests\Rules\UnusedPublicClassMethodRule\UnusedPublicClassMethodRuleTest::testRule with data set #18
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'
+'08: Public method "TomasVotruba\UnusedPublic\Tests\Rules\UnusedPublicClassMethodRule\Source\Caller1::callIt()" is never used
+    💡 Either reduce visibility or annotate it or its class with @api
+08: Public method "TomasVotruba\UnusedPublic\Tests\Rules\UnusedPublicClassMethodRule\Source\Caller2::callIt()" is never used
+    💡 Either reduce visibility or annotate it or its class with @api
 '
 ```

**Comment by staabm** [comment]
ready

File: composer.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "keywords": ["static analysis", "phpstan-extension"],
     "require": {
         "php": "^8.1",
-        "phpstan/phpstan": "^1.9.3",
+        "phpstan/phpstan": "^1.10.0",
```
Comment (staabm) [inline composer.json:9]: `$callerType->getReferencedClasses()` was added in 1.10.0

File: src/Collectors/MethodCallCollector.php [hunk 2]
```diff
@@ -56,22 +56,25 @@ public function processNode(Node $node, Scope $scope): ?array
...
-        $classMethodCallReference = $this->classMethodCallReferenceResolver->resolve($node, $scope);
-        if (! $classMethodCallReference instanceof MethodCallReference) {
-            return null;
-        }
+        $classMethodReferences = [];
 
-        $className = $classMethodCallReference->getClass();
-        $methodName = $classMethodCallReference->getMethod();
+        $classMethodCallReferences = $this->classMethodCallReferenceResolver->resolve($node, $scope);
+        foreach($classMethodCallReferences as $classMethodCallReference) {
```
Comment (staabm) [inline src/Collectors/MethodCallCollector.php]: just put a loop around the pre-existing checks to handle the newly returned iterable type from `resolve`

**Comment by staabm** [comment]
@TomasVotruba any opinon on this one?

(the rector build error seems unrelated)

**Comment by TomasVotruba** [comment]
Looks very good :+1:  What exactly do you need help with? Seems tests are failing.

I'll check the Rector job, it should not even run.

**Comment by staabm** [comment]
no help needed, only a merge button pusher :).

> Seems tests are failing.

fixed the test-expectation which went invalid because the cs-fixer removed a unused import.

**Comment by TomasVotruba** [comment]
CI must pass before merge ;) 

Could you rebase the branch? Rector should become silent

**Comment by staabm** [comment]
looks better now, thank you

**Comment by staabm** [comment]
friendly ping @TomasVotruba  :)

**Comment by TomasVotruba** [comment]
Looks good, thank you :clap:

**Pull request merged**